UnitTests/ControllerTests/EPubsControllerTest.cs

```diff
@@ -0,0 +1,216 @@
+                SourceType = ResourceType.Media
+            };
+
+            var postResult = await _controller.PostEPub(ePub);
+            var createdResult = postResult.Result as CreatedAtActionResult;
+
+            Assert.NotNull(createdResult);
+
+            var deleteResult = await _controller.DeleteEPub(ePub.Id);
+
+            Assert.NotNull(deleteResult.Value);
+            Assert.Equal(ePub, deleteResult.Value);
+
+            var getResult = await _controller.GetEPub(ePub.Id);
+
+            Assert.Equal(Status.Deleted, getResult.Value.IsDeletedStatus);
+        }
+
+        [Fact]
+        public async Task Get_EPubs_By_Source()
+        {
+            var ePubs = new List<EPub> {
+                new EPub
+                {
+                    Title = "example1",
+                    Language = "en-US",
+                    Filename = "filename_example",
+                    Author = "author_example",
+                    Publisher = "publisher_example",
+                    SourceId = "mediaId_example",
+                    SourceType = ResourceType.Media
+                },
+                new EPub
+                {
+                    Title = "example2",
+                    Language = "en-US",
+                    Filename = "filename_example",
+                    Author = "author_example",
+                    Publisher = "publisher_example",
+                    SourceId = "courseId_example",
+                    SourceType = ResourceType.Course
+                },
+                new EPub
+                {
+                    Title = "example3",
+                    Language = "en-US",
+                    Filename = "filename_example",
+                    Author = "author_example",
+                    Publisher = "publisher_example",
+                    SourceId = "mediaId_example",
+                    SourceType = ResourceType.Media
+                }
+            };
+
+            foreach (var ePub in ePubs)
+            {
+                var postResult = await _controller.PostEPub(ePub);
+                var createdResult = postResult.Result as CreatedAtActionResult;
+
+                Assert.NotNull(createdResult);
+            }
+
+            var getResult = await _controller.GetEPubsBySource(ResourceType.Media.ToString(), ePubs[0].SourceId);
+            List<EPub> ePubsBySource = getResult.Value.ToList();
+
+            Assert.Equal(2, ePubsBySource.Count);
+            Assert.Equal(ePubs[0], ePubsBySource[0]);
+            Assert.Equal(ePubs[2], ePubsBySource[1]);
+        }
+
+        [Fact]
+        public async Task Post_Invalid_and_Valid_EPubs()
+        {
+            var ePub = new EPub
+            {
+                Title = "example0",
+                Language = "en-US"
+            };
+
+            var postResult = await _controller.PostEPub(ePub);
+            var badRequestResult = postResult.Result as BadRequestObjectResult;
+
+            Assert.NotNull(badRequestResult);
```

Comment:
If this `PostEPub` were to unexpectedly succeed, would `badRequestResult` be null? Just trying to understand how these `as *Result` work.

Reply:
Exactly, `badRequestResult` would be null if the POST succeeded. This is because the postResult.Result would be a `CreatedAtActionResult` instead of a `BadRequestObjectResult`, so the `as` operator would fail and result in null.

From Microsoft's docs on the `as` operator (https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/type-testing-and-cast#as-operator):
> The as operator explicitly converts the result of an expression to a given reference or nullable value type. If the conversion is not possible, the as operator returns null. Unlike a cast expression, the as operator never throws an exception.